Src/F_Interfaces/Particle/AMReX_particlecontainer_mod.F90

```diff
@@ -0,0 +1,135 @@
+module amrex_particlecontainer_module
+
+  use iso_c_binding
+  use amrex_base_module
+  use amrex_string_module
+  use amrex_fort_module, only: amrex_particle_real
+  
+  implicit none
+
+  private
+
+  ! public routines
+  public :: amrex_particlecontainer_init, amrex_particlecontainer_finalize
+  public :: amrex_init_particles_one_per_cell, amrex_write_particles, amrex_particle_redistribute
+  public :: amrex_get_particles, amrex_num_particles
+  
+  type, bind(C), public :: amrex_particle
+     real(amrex_particle_real)    :: pos(AMREX_SPACEDIM) !< Position
+     real(amrex_particle_real)    :: vel(AMREX_SPACEDIM) !< Particle velocity
+     integer(c_int)               :: id
+     integer(c_int)               :: cpu
+  end type amrex_particle
+
+  interface
+     subroutine amrex_fi_new_particlecontainer (pc,amrcore) bind(c)
+       import
+       implicit none
+       type(c_ptr) :: pc
+       type(c_ptr), value :: amrcore
+     end subroutine amrex_fi_new_particlecontainer
+
+     subroutine amrex_fi_delete_particlecontainer (pc) bind(c)
+       import
+       implicit none
+       type(c_ptr), value :: pc
+     end subroutine amrex_fi_delete_particlecontainer
+
+     subroutine amrex_fi_init_particles_one_per_cell (pc) bind(c)
+       import
+       implicit none
+       type(c_ptr), value :: pc
+     end subroutine amrex_fi_init_particles_one_per_cell
+
+     subroutine amrex_fi_write_particles (pc, dirname, pname, is_checkpoint) bind(c)
+       import
+       implicit none
+       type(c_ptr), value :: pc
+       character(kind=c_char), intent(in) :: dirname(*)
+       character(kind=c_char), intent(in) :: pname(*)
+       logical, intent(in)                :: is_checkpoint
```

Comment:
(1) This needs to be value.  (2) logical is not necessary the same type as C++ bool.  For portability, we need to use `integer(c_int)` and `int`.  (Note that we still want to use `logical` for Fortran function argument, but we do a conversion to `integer(c_int)` before the call to `amrex_fi_write_particles`.